lib/utils/ast-utils/extract-expression-references.ts

```diff
@@ -0,0 +1,312 @@
+import type { Rule } from "eslint"
+import type { Variable } from "eslint-scope"
+import type {
+    ArrayPattern,
+    ArrowFunctionExpression,
+    CallExpression,
+    Expression,
+    ForOfStatement,
+    FunctionDeclaration,
+    FunctionExpression,
+    Identifier,
+    MemberExpression,
+    ObjectPattern,
+    Pattern,
+} from "estree"
+import { findVariable, getParent } from "."
+
+export type ExpressionReference =
+    | {
+          // The result of the expression is not referenced.
+          type: "unused"
+          node: Expression
+      }
+    | {
+          // Unknown what the expression was referenced for.
+          type: "unknown"
+          node: Expression
+      }
+    | {
+          // The expression is exported.
+          type: "exported"
+          node: Expression
+      }
+    | {
+          // The expression is referenced for member access.
+          type: "member"
+          node: Expression
+          memberExpression: MemberExpression
+      }
+    | {
+          // The expression is referenced for destructuring.
+          type: "destructuring"
+          node: Expression
+          pattern: ObjectPattern | ArrayPattern
+      }
+    | {
+          // The expression is referenced to give as an argument.
+          type: "argument"
+          node: Expression
+          callExpression: CallExpression
+      }
+    | {
+          // The expression is referenced to call.
+          type: "call"
+          node: Expression
+      }
+    | {
+          // The expression is referenced for iteration.
+          type: "iteration"
+          node: Expression
+          for: ForOfStatement
+      }
+
+type AlreadyChecked = {
+    variables: Set<Variable>
+    functions: Map<
+        FunctionDeclaration | FunctionExpression | ArrowFunctionExpression,
+        Set<number>
+    >
+}
+/** Extract references from the given expression */
+export function* extractExpressionReferences(
+    node: Expression,
+    context: Rule.RuleContext,
+): IterableIterator<ExpressionReference> {
```

Comment:
```suggestion
): Iterable<ExpressionReference> {
```

A little suggestion here. It's correct to return an `IterableIterator` here but this might make it harder to change the function in the future. If we ever re-implement `extractExpressionReferences` as a normal function, it will be hard to guarantee the `IterableIterator` interface. Plus, only the `Iterable` part of the `IterableIterator` interface is typically used.

I avoid the `IterableIterator` in my project for those reasons. If I really need an iterator, I can always just all `iterable[Symbol.iterator]()` anyway.

Reply:
Thank you for your review!
I hadn't really thought about the difference between them. I will change this PR.